---
title: "Interactive Analysis with GeneTonic"
author: "tucca-rna-seq"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: united
---

# Introduction

This RMarkdown document is a playground for interactive exploration of RNA-seq
analysis results from the `tucca-rna-seq` pipeline. It uses the `GeneTonic`
package to visualize and interpret differential expression and enrichment
analysis results.

Before you begin, make sure you have successfully completed a `tucca-rna-seq`
run. The objects generated by the pipeline are required for this analysis.

# Setup

## Load Libraries

First, we load all the necessary R packages.

```{r setup, message=FALSE, warning=FALSE}
# Restore the environment from the lockfile to ensure reproducibility.
# This will install any missing packages from the renv.lock file.
renv::restore()

# All functions in this notebook will be called with explicit package namespaces
# (e.g., `GeneTonic::GeneTonic()`) to ensure clarity and avoid conflicts.
# This makes `library()` calls unnecessary.
```

# Configuration

Here, you need to specify which analysis and contrast you want to explore.
These names correspond to the configurations in your `config.yaml`.

```{r configuration}
# --- USER INPUT REQUIRED ---
# Replace with the 'analysis_name' from your config.yaml
analysis_name <- "my_analysis"
# Replace with the 'contrast_name' from your config.yaml
contrast_name <- "condition_B_vs_A"

# You can find available analyses and contrasts by looking at the directory
# names in `../tucca-rna-seq/resources/deseq2/`
```

# Load Data

Now, we load the data generated by the `tucca-rna-seq` pipeline based on the configuration above.

```{r load_data}
# Construct file paths
dds_path <- base::file.path("..", "resources", "deseq2", analysis_name, "dds.RDS")
res_path <- base::file.path("..", "resources", "deseq2", analysis_name, contrast_name, "wald.RDS")
ora_path <- base::file.path("..", "resources", "enrichment", analysis_name, contrast_name, "ora_results.RDS")
gsea_path <- base::file.path("..", "resources", "enrichment", analysis_name, contrast_name, "gsea_results.RDS")
config_path <- base::file.path("..", "config", "config.yaml")

# Load the main objects
dds <- base::readRDS(dds_path)
res_de <- base::readRDS(res_path) # This is the DESeqResults object
ora_results <- base::readRDS(ora_path)
gsea_results <- base::readRDS(gsea_path)
config <- yaml::read_yaml(config_path)

# The results objects are lists (e.g., for GO and KEGG).
# We will select one for the GeneTonic visualization.
res_enrich <- ora_results$GO # or ora_results$KEGG, or gsea_results$GO, etc.
```

# Prepare Inputs for GeneTonic

`GeneTonic` requires a few specific inputs: `dds`, `res_de`, `res_enrich`, and
`annotation_obj`. We have the first three, now we create the annotation object.

## Create `annotation_obj`

The `annotation_obj` is a data frame that maps gene IDs to gene names. We'll
create it using the correct organism annotation package.

```{r create_annotation_obj}
# Get OrgDb package name from config.
# This logic mirrors the pipeline scripts to correctly identify the package
# name, whether it was installed from Bioconductor or built locally.
enrichment_params <- config$enrichment
install_method <- enrichment_params$install_method
install_source <- enrichment_params$install_source

# Determine the actual package name
org_db_pkg <- if (install_method == "local") {
  # When built locally, find the package name from the build directory.
  # The path in config.yaml is relative to the project root.
  local_build_path <- base::file.path("..", install_source)
  pkg_dirs <- base::list.dirs(path = local_build_path, full.names = FALSE, recursive = FALSE)
  pkg_name <- pkg_dirs[grepl("^org\\..+\\.db$", pkg_dirs)]
  if (length(pkg_name) == 0) {
    base::stop("Could not find a locally built OrgDb package in the specified directory: ", local_build_path)
  }
  pkg_name[1]
} else {
  enrichment_params$org_db_pkg
}

base::message("Attempting to load OrgDb package: ", org_db_pkg)

if (!base::require(org_db_pkg, character.only = TRUE)) {
  base::message("OrgDb package not found or could not be loaded. Please ensure the pipeline has run successfully and that you are in the correct R environment.")
  annotation_obj <- NULL
} else {
  # The annotation object needs columns: 'gene_id', 'gene_name'
  # We will map from ENSEMBL IDs (if they are the rownames of our dds)
  keys <- utils::head(base::rownames(dds))
  keytype <- "ENSEMBL" # Change if your dds uses other identifiers like ENTREZID

  base::message(base::paste("Using", org_db_pkg, "for annotation."))

  annotation_obj <- AnnotationDbi::select(
    base::get(org_db_pkg),
    keys = AnnotationDbi::keys(base::get(org_db_pkg), keytype = keytype),
    columns = c("SYMBOL"),
    keytype = keytype
  ) %>%
    dplyr::rename(gene_id = dplyr::all_of(keytype), gene_name = "SYMBOL") %>%
    dplyr::filter(!base::is.na(gene_name))
}
```

## "Shake" the enrichment results

`GeneTonic` has `shake_` functions to convert standard enrichment result objects into the required `data.frame` format.

```{r shake_results}
# We will use the 'shaken' version of the enrichment results
# This function is smart enough to handle both gseaResult and enrichResult
res_enrich_shaken <- GeneTonic::shake_topGOtableResult(res_enrich)
```

# Launch GeneTonic

With all the ingredients ready, we can now create the `GeneTonicList` object and launch the interactive application.

```{r launch_gene_tonic, eval=interactive()}
if (!base::is.null(res_enrich) && !base::is.null(annotation_obj)) {
  # Create the GeneTonicList object
  gtl <- GeneTonic::GeneTonicList(
    dds = dds,
    res_de = res_de,
    res_enrich = res_enrich_shaken,
    annotation_obj = annotation_obj
  )

  # Launch the app!
  GeneTonic::GeneTonic(gtl = gtl, project_id = "My Project")
} else {
  base::message("Could not launch GeneTonic because enrichment results or annotation object are missing.")
}
```

# Reporting with `happy_hour`

`GeneTonic` also provides a function called `happy_hour()` to create a non-interactive HTML report summarizing key results. This is useful for sharing your findings.

You can select genes and gene sets of interest to focus the report.

```{r happy_hour, eval=FALSE}
# Example of how to run happy_hour - uncomment and customize to run
# my_genesets <- utils::head(res_enrich_shaken$gs_id, 5)
# my_genes <- base::c("GENE1_ID", "GENE2_ID") # Replace with gene IDs of interest
#
# GeneTonic::happy_hour(
#   gtl = gtl,
#   project_id = "MyReport",
#   mygenesets = my_genesets,
#   mygenes = my_genes,
#   open_after_creating = TRUE
# )
``` 